File: lib/scuti/model/task.ex
# Copyright 2023 Clivern. All rights reserved.
# Use of this source code is governed by the MIT
# license that can be found in the LICENSE file.

defmodule Scuti.Model.Task do
  @moduledoc """
  Task Model
  """

  use Ecto.Schema
  import Ecto.Changeset

  schema "tasks" do
    field :uuid, Ecto.UUID
    field :deployment_id, :id
    field :payload, :string
    field :result, :string
    field :status, :string
    field :run_at, :utc_datetime

    timestamps()
  end

  @doc false
  def changeset(task, attrs) do
    task
    |> cast(attrs, [
      :uuid,
      :deployment_id,
      :payload,
      :result,
      :status,
      :run_at
    ])
    |> validate_required([
      :uuid,
      :deployment_id,
      :payload,
      :result,
      :status,
      :run_at
    ])
  end
end

File: agent/core/controller/task.go
// Copyright 2023 Clivern. All rights reserved.
// Use of this source code is governed by the MIT
// license that can be found in the LICENSE file.

package controller

import (
	"fmt"
	"net/http"

	"github.com/labstack/echo/v4"
	log "github.com/sirupsen/logrus"
)

// GetTaskAction Controller
func GetTaskAction(c echo.Context, globalContext *GlobalContext) error {

	log.Info(`Incoming Request to GetTask Action`)

	uuid := c.Param("uuid")

	task := globalContext.GetDatabase().GetTaskByUUID(uuid)

	if task.IsNil() {
		return c.JSON(http.StatusNotFound, map[string]interface{}{
			"errorMessage": fmt.Sprintf("Task with UUID %s not found", uuid),
		})
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"id":        task.ID,
		"uuid":      task.UUID,
		"did":       task.DID,
		"payload":   task.Payload,
		"result":    task.Result,
		"status":    task.Status,
		"meta":      task.Meta,
		"createdAt": task.CreatedAt,
		"updatedAt": task.UpdatedAt,
	})
}

// DeleteTaskAction Controller
func DeleteTaskAction(c echo.Context, globalContext *GlobalContext) error {

	log.Info(`Incoming Request to DeleteTask Action`)

	uuid := c.Param("uuid")

	task := globalContext.GetDatabase().GetTaskByUUID(uuid)

	if task.IsNil() {
		return c.JSON(http.StatusNotFound, map[string]interface{}{
			"errorMessage": fmt.Sprintf("Task with UUID %s not found", uuid),
		})
	}

	globalContext.GetDatabase().DeleteTaskByUUID(uuid)

	return c.NoContent(http.StatusNoContent)
}
